### ISSUE TITLE
Improve End2End ServerTryCancel tests.. 

### ISSUE THREAD

**Issue by sreecha** [grpc/grpc#5520]
For some background, read the conversation at: https://github.com/grpc/grpc/pull/5493

The summary is that we need a better way test cases where the RPC is called (by server calling `TryCancel`) WHILE the RPC is being processed by server or the client. More specifically, the following:

`File: test/cpp/end2end/end2end_test.cc`
```
void TestRequestStreamServerCancel(...) {
 ...
  case CANCEL_DURING_PROCESSING:
        // If the RPC is cancelled by server before / during messages from the
        // client, it means that the client most likely did not get a chance to
        // send all the messages it wanted to send. i.e num_msgs_sent <=
        // num_msgs_to_send
        EXPECT_LE(num_msgs_sent, num_msgs_to_send);  //
        break;
...
}

void TestResponseStreamServerCancel(...) {
...
      case CANCEL_DURING_PROCESSING:
        // Server cancelled while writing messages. Client must have read less
        // than or equal to the expected number of messages
        EXPECT_LE(num_msgs_read, kNumResponseStreamsMsgs);
        break;
...
}

void TestBidiStreamServerCancel(...) {
...
      case CANCEL_DURING_PROCESSING:
        EXPECT_LE(num_msgs_sent, num_messages);
        EXPECT_LE(num_msgs_read, num_msgs_sent);
        break;

      case CANCEL_AFTER_PROCESSING:
        EXPECT_EQ(num_msgs_sent, num_messages);
        // The Server cancelled after reading the last message and after writing
        // the message to the client. However, the RPC cancellation might have
        // taken effect before the client actually read the response.
        EXPECT_LE(num_msgs_read, num_msgs_sent);
        break;
...
}

```
In all the above, the check EXPECT_LE() is redundant as it is going to be true no matter what.. 
The main scenario that the snippets above are testing is the cases where RPC was cancelled in parallel while server (or client) are processing the requests/responses....

We need a better way to test the above.. Something that can verify that the RPC cancellation was issued in parallel while server and/or client are in the middle of processing the requests/responses

(Note: The fact that the RPC is successfully cancelled is being checked at the very end.)

**Comment by pgrosu** [comment]
Hi Sree,

Hope this might bring us closer to solving this problem.  I should mention that I'm still reading through all the core code to become familiar with the whole implementation.  So you have the following function in [core/surface/call.c](https://github.com/grpc/grpc/blob/3f19f427714c08d1fce449cf8904fa1494602285/src/core/surface/call.c#L676-L692)

```C
grpc_call_error grpc_call_cancel_with_status(grpc_call *c,
                                             grpc_status_code status,
                                             const char *description,
                                             void *reserved) { ... }
```

It is possible for a `RST_STREAM` frame - as noted in the [HTTP2 spec](https://http2.github.io/http2-spec/#RST_STREAM) - to be sent with a request for cancellation or to indicate that a [CANCEL (0x8)](https://http2.github.io/http2-spec/#ErrorCodes) error occurred.  And it mentions that:

> However, after sending the RST_STREAM, the sending endpoint MUST be prepared to receive and process additional frames sent on the stream that might have been sent by the peer prior to the arrival of the RST_STREAM. 

This would allow the client to check if the stream was canceled by the server.

In fact for a STREAM_CLOSED (0x5) error, it is possible to perform the following [Stream Error Handling](https://http2.github.io/http2-spec/#StreamErrorHandler):

> Normally, an endpoint SHOULD NOT send more than one RST_STREAM frame for any stream. However, an endpoint MAY send additional RST_STREAM frames if it receives frames on a closed stream after more than a round-trip time. This behavior is permitted to deal with misbehaving implementations. 

Would this not solve the issue?

Let me know what you think.

Thanks,
Paul

**Comment by sreecha** [comment]
Hi Paul,
Thanks for looking into this.

The issue is not about checking whether cancellation was received by the client or not. The fact that the cancellation was seen on the server side and the client side is good enough - and the current tests do check those conditions.

The reason I opened this issue was because for a subset of these tests, the cancellation happens _somewhere in the middle_ (which is totally random) of servers/clients reading/writing messages depending on the test case. Currently the test verification checks if the RPC was cancelled from both client and server perspective - but we do not have anything that verifies the  "_cancelled somewhere in the middle_" aspect of the tests.

At this point, I can't even define what it means to verify that aspect :) . I tried verifying that the number of messages successfully sent or received are less than or equal to the expected number of messages to send or receive (depending on the test) but then again, considering the way these tests are written, those conditions are is going to be true no matter what!. So we need a different way to assert this.  

The more I think of this, the more I feel it is low priority :).

**Comment by pgrosu** [comment]
Hi Sree,

No problem, but could you point me to a specific portion of the code where you simulating this cancel that is not originating at either the server or the client.  Usually even if you are running things via [Quic in Chromium](https://chromium.googlesource.com/chromium/src/net/+/27966cd267379b66a1baeb2b12bfd8950785464f/quic/quic_spdy_server_stream.cc), there is a Session ID and a Stream ID to work with which would basically in our case be part of a channel:

```C++
QuicSpdyServerStream::QuicSpdyServerStream(QuicStreamId id,
                                           QuicSession* session)
```

In fact you can have a dispatcher or a broker to coordinate via.  For more info on Quic below is a link in case it might help here:

https://www.chromium.org/quic

https://chromium.googlesource.com/chromium/src/net/+/refs/heads/master/quic/

~p

**Comment by sreecha** [comment]
>  portion of the code where you simulating this cancel that is not originating at either the server or the client
I am not simulating a cancel. The cancel originates from Server via `TryCancel` call on the server (see the function `ServerTryCancel` in `test/cpp/end2end/end2end_test.cc` and `test/cpp/end2end/test_service_impl.cc`

**Comment by sreecha** [comment]
I am closing this. Can't think of any better way to improve the tests.
